cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/polling/task/ClouderaManagerDeployClientConfigListenerTask.java

```diff
@@ -66,6 +64,44 @@ public boolean checkStatus(ClouderaManagerCommandPollerObject clouderaManagerPol
         }
     }
 
+    private List<ClouderaManagerProduct> getClouderaManagerProductsFromStack(Stack stack) {
+        return Optional.ofNullable(stack).map(Stack::getCluster).map(Cluster::getComponents)
+                .orElse(Set.of())
+                .stream()
+                .filter(clusterComponent -> ComponentType.CDH_PRODUCT_DETAILS.equals(clusterComponent.getComponentType()))
+                .map(ClusterComponent::getAttributes)
+                .map(toAttributeClass(ClouderaManagerProduct.class))
+                .collect(Collectors.toList());
+    }
+
+    private ApiParcelList getClouderaManagerParcels(ApiClient apiClient, String stackName) throws ApiException {
+        ParcelsResourceApi parcelsResourceApi = clouderaManagerClientFactory.getParcelsResourceApi(apiClient);
+        return parcelsResourceApi.readParcels(stackName, "summary");
+    }
+
+    private List<ApiParcel> getNotActivatedOrMissingParcels(List<ClouderaManagerProduct> products, ApiParcelList parcels) {
+        List<ApiParcel> notActivated = new ArrayList<>(parcels.getItems().size());
+        products.forEach(product -> {
+            parcels.getItems()
+                    .stream()
+                    .filter(parcel -> product.getName().equals(parcel.getProduct())
```

Comment:
just a note. I would extract this – and the `ifPresentOrElse` – logic to a separate method to make the whole stream chain more readable

Reply:
made more clean.